programs/keeper/Keeper.cpp

```diff
@@ -0,0 +1,474 @@
+{
+    if (config().hasOption("help"))
+    {
+        Poco::Util::HelpFormatter help_formatter(Keeper::options());
+        auto header_str = fmt::format("{} [OPTION] [-- [ARG]...]\n"
+                                      "positional arguments can be used to rewrite config.xml properties, for example, --http_port=8010",
+                                      commandName());
+        help_formatter.setHeader(header_str);
+        help_formatter.format(std::cout);
+        return 0;
+    }
+    if (config().hasOption("version"))
+    {
+        std::cout << DBMS_NAME << " keeper version " << VERSION_STRING << VERSION_OFFICIAL << "." << std::endl;
+        return 0;
+    }
+
+    return Application::run(); // NOLINT
+}
+
+void Keeper::initialize(Poco::Util::Application & self)
+{
+    BaseDaemon::initialize(self);
+    logger().information("starting up");
+
+    LOG_INFO(&logger(), "OS Name = {}, OS Version = {}, OS Architecture = {}",
+        Poco::Environment::osName(),
+        Poco::Environment::osVersion(),
+        Poco::Environment::osArchitecture());
+}
+
+std::string Keeper::getDefaultConfigFileName() const
+{
+    return "keeper_config.xml";
+}
+
+void Keeper::defineOptions(Poco::Util::OptionSet & options)
+{
+    options.addOption(
+        Poco::Util::Option("help", "h", "show help and exit")
+            .required(false)
+            .repeatable(false)
+            .binding("help"));
+    options.addOption(
+        Poco::Util::Option("version", "V", "show version and exit")
+            .required(false)
+            .repeatable(false)
+            .binding("version"));
+    BaseDaemon::defineOptions(options);
+}
+
+int Keeper::main(const std::vector<std::string> & /*args*/)
+{
+    Poco::Logger * log = &logger();
+
+    UseSSL use_ssl;
+
+    MainThreadStatus::getInstance();
+
+#if !defined(NDEBUG) || !defined(__OPTIMIZE__)
+    LOG_WARNING(log, "Keeper was built in debug mode. It will work slowly.");
+#endif
+
+#if defined(SANITIZER)
+    LOG_WARNING(log, "Keeper was built with sanitizer. It will work slowly.");
+#endif
+
+    auto shared_context = Context::createShared();
+    global_context = Context::createGlobal(shared_context.get());
+
+    global_context->makeGlobalContext();
+    global_context->setApplicationType(Context::ApplicationType::SERVER);
+
+    if (!config().has("keeper_server"))
+        throw Exception(ErrorCodes::NO_ELEMENTS_IN_CONFIG, "Keeper configuration (<keeper_server> section) not found in config");
+
+
+    std::string path;
+
+    if (config().has("keeper_server.storage_path"))
+        path = config().getString("keeper_server.storage_path");
+    else if (config().has("keeper_server.log_storage_path"))
+        path = config().getString("keeper_server.log_storage_path");
+    else if (config().has("keeper_server.snapshot_storage_path"))
+        path = config().getString("keeper_server.snapshot_storage_path");
+    else
+        path = std::filesystem::path{DBMS_DEFAULT_PATH} / "coordination/logs";
```

Comment:
Is it still possible to use embedded keeper in clickhouse-server? Do we need some protection from running multiple keepers (for example, embedded and standalone ones) in one directory?

Reply:
It just shouldn't start because of the same ports or the same PID files. But it's always possible to turn your data into the garbage. It's also possible with `clickhouse-server`. Maybe it would be better to have a separate path, I'll add `KEEPER_DEFAULT_PATH`.